pkg/initializer/validator/datavalidator.go

```diff
@@ -250,17 +265,73 @@ func (d *DataValidator) checkRevisionConsistency(etcdRevision, failBelowRevision
 		d.Logger.Infof("No snapshot found.")
 		if etcdRevision < failBelowRevision {
 			d.Logger.Infof("current etcd revision (%d) is less than fail below revision (%d): possible data loss", etcdRevision, failBelowRevision)
-			return FailBelowRevisionConsistencyError, nil
+			return FailBelowRevisionConsistencyError, latestSnapshotRevision, nil
 		}
-		return DataDirectoryValid, nil
+		return DataDirectoryValid, latestSnapshotRevision, nil
 	}
 
 	if etcdRevision < latestSnapshotRevision {
-		d.Logger.Infof("current etcd revision (%d) is less than latest snapshot revision (%d): possible data loss", etcdRevision, latestSnapshotRevision)
+		d.Logger.Infof("current etcd revision (%d) is less than latest snapshot revision (%d)", etcdRevision, latestSnapshotRevision)
+		return EtcdRevisionConsistencyError, latestSnapshotRevision, nil
+	}
+
+	return DataDirectoryValid, latestSnapshotRevision, nil
+}
+
+// checkFullRevisionConsistency starts an embedded etcd and then compares the latest revision of etcd db file and the latest snapshot revision to verify that the etcd revision is not lesser than snapshot revision.
+// Return DataDirStatus indicating whether  WALs file have uncommited data which it was unable to flush to DB or latest DB revision is still less than snapshot revision.
+func (d *DataValidator) checkFullRevisionConsistency(dataDir string, latestSnapshotRevision int64) (DataDirStatus, error) {
+
+	var latestSyncedEtcdRevision int64
+	d.Logger.Info("Starting embedded etcd server...")
+
+	e, err := restorer.StartEmbeddedEtcd(logrus.NewEntry(d.Logger), dataDir, d.Config.EmbeddedEtcdQuotaBytes)
+	if err != nil {
+		d.Logger.Infof("unable to start embedded etcd: %v", err)
+		return DataDirectoryCorrupt, err
+	}
+	defer func() {
+		e.Server.Stop()
+		e.Close()
+	}()
+	client, err := clientv3.NewFromURL(e.Clients[0].Addr().String())
+	if err != nil {
+		d.Logger.Infof("unable to get the embedded etcd client: %v", err)
+		return DataDirectoryCorrupt, err
+	}
+
+	start := time.Now()
+	timer := time.NewTimer(60 * time.Second)
+	for {
```

Comment:
How about simplifying the loop like this?
```go
       defer timer.Stop()
       waitLoop: for {
		select {
		case <-timer.C:
			break waitLoop
		default:
			latestSyncedEtcdRevision, _ = getLatestSyncedRevision(client)
			if latestSyncedEtcdRevision >= latestSnapshotRevision {
				d.Logger.Infof("After starting embeddedEtcd backend DB file revision (%d) is greater than or equal to latest snapshot revision (%d): no data loss", latestSyncedEtcdRevision, latestSnapshotRevision)
				break waitLoop
			}
			time.Sleep(1 * time.Second)
		}
	}
```

Reply:
I have made all changes which you suggested 😊